tensorflow/python/ops/linalg_grad.py

```diff
@@ -352,32 +354,45 @@ def _SvdGrad(op, grad_s, grad_u, grad_v):
             array_ops.expand_dims(s2, -2) - array_ops.expand_dims(s2, -1)),
         array_ops.zeros_like(s))
     s_inv_mat = array_ops.matrix_diag(math_ops.reciprocal(s))
+
+    v1 = v[..., :, :m]
+    grad_v1 = grad_v[..., :, :m]
+
     u_gu = math_ops.matmul(u, grad_u, adjoint_a=True)
-    v_gv = math_ops.matmul(v, grad_v, adjoint_a=True)
+    v_gv = math_ops.matmul(v1, grad_v1, adjoint_a=True)
 
-    if m == n:
-      f_u = f * u_gu
-      f_v = f * v_gv
-    else:
-      dv2 = array_ops.matrix_transpose(v_gv[..., m:n, :m]) - v_gv[..., :m, m:n]
```

Comment:
I should explain what happened to `dv2`.  `dv2` has two terms:
1. `array_ops.matrix_transpose(v_gv[..., m:n, :m])` and
2. `-v_gv[..., :m, m:n]`.

When `grad_a_nouv` gets right-multiplied by the adjoint of `v`, the submatrix `dv2` gets right-multiplied by the adjoint of `v[..., :, m:n]`.  After right-multiplication by `v[..., :, m:n]`,
1. the first term of `dv2` becomes the first term of `term2_nous`: `math_ops.matmul(grad_v1, proj_v1_perp, adjoint_a=True)`
2. the second term of `dv2` becomes the second term of `term2_nous`: `-math_ops.matmul(v1t_gv2, v2, adjoint_b=True)`

To see why the first terms match, use the identity
```
math_ops.matmul(v[..., :, :m], v[..., :, :m], adjoint_b=True)
   + math_ops.matmul(v[..., :, m:n], v[..., :, m:n], adjoint_b=True)
       == linalg_ops.eye(n, dtype=v.dtype)
```
which follows from the fact that `v` is unitary.

My motivation for separating the terms of `dv2` was to make clear what the contribution to `grad_a` is from the "redundant" columns of V when `full_matrices=True`.

If the math here isn't clear, just let me know and I can write it down and upload it.